src/modules/job-info/job-info.c

```diff
@@ -0,0 +1,420 @@
+        flux_future_destroy (ctx->f);
+        free (ctx);
+    }
+}
+
+static void lookup_continuation (flux_future_t *f, void *arg);
+
+struct lookup_ctx *lookup_ctx_create (struct info_ctx *ctx,
+                                      const flux_msg_t *msg,
+                                      flux_jobid_t id,
+                                      int flags)
+{
+    struct lookup_ctx *l = calloc (1, sizeof (*l));
+    int saved_errno;
+
+    if (!l)
+        return NULL;
+
+    l->ctx = ctx;
+    l->id = id;
+    l->flags = flags;
+    l->active = true;
+
+    if (l->flags & FLUX_JOB_INFO_WATCH) {
+        l->lookup_flags |= FLUX_KVS_WATCH;
+        l->lookup_flags |= FLUX_KVS_WATCH_APPEND;
+    }
+
+    if (!(l->msg = flux_msg_copy (msg, true))) {
+        flux_log_error (ctx->h, "%s: flux_msg_copy", __FUNCTION__);
+        goto error;
+    }
+
+    return l;
+
+error:
+    saved_errno = errno;
+    lookup_ctx_destroy (l);
+    errno = saved_errno;
+    return NULL;
+}
+
+/* 'pp' is an in/out parameter pointing to input buffer.
+ * Set 'tok' to next \n-terminated token, and 'toklen' to its length.
+ * Advance 'pp' past token.  Returns false when input is exhausted.
+ */
+static bool info_parse_next (const char **pp, const char **tok,
+                                 size_t *toklen)
+{
+    char *term;
+
+    if (!(term = strchr (*pp, '\n')))
+        return false;
+    *tok = *pp;
+    *toklen = term - *pp + 1;
+    *pp = term + 1;
+    return true;
+}
+
+static int lookup_key (struct lookup_ctx *l)
+{
+    char key[64];
+
+    if (l->f) {
+        flux_future_destroy (l->f);
+        l->f = NULL;
+    }
+
+    if (flux_job_kvs_key (key, sizeof (key), l->active, l->id, "eventlog") < 0) {
+        flux_log_error (l->ctx->h, "%s: flux_job_kvs_key", __FUNCTION__);
+        return -1;
+    }
+
+    if (!(l->f = flux_kvs_lookup (l->ctx->h, NULL, l->lookup_flags, key))) {
+        flux_log_error (l->ctx->h, "%s: flux_kvs_lookup", __FUNCTION__);
+        return -1;
+    }
+
+    if (flux_future_then (l->f, -1, lookup_continuation, l) < 0) {
+        flux_log_error (l->ctx->h, "%s: flux_future_then", __FUNCTION__);
+        return -1;
+    }
+
+    return 0;
+}
+
+static void lookup_continuation (flux_future_t *f, void *arg)
+{
+    struct lookup_ctx *l = arg;
+    struct info_ctx *ctx = l->ctx;
+    const char *s;
+
+    if (flux_kvs_lookup_get (f, &s) < 0) {
+        if (errno == ENOENT && l->active) {
+            /* transition / try the inactive key */
+            l->active = false;
+            if (lookup_key (l) < 0)
+                goto error;
+            return;
+        }
+        else if (errno == ENODATA && (l->flags & FLUX_JOB_INFO_WATCH)) {
+            if (flux_respond_error (ctx->h, l->msg, ENODATA, NULL) < 0)
+                flux_log_error (ctx->h, "%s: flux_respond_error", __FUNCTION__);
+            goto done;
+        }
+        else
+            flux_log_error (ctx->h, "%s: flux_kvs_lookup_get", __FUNCTION__);
```

Comment:
Maybe this error should be supressed (at least for ENOENT), since it is likely a result of looking up an invalid jobid? (No reason to spam the logs, since the error is returned to the user anyway)